[Model] Support telechat2

shunxing12345 opened the pull request:
Related #5776 

FIX #6503

Commit:
support telechat2

shunxing12345 commented:
Title: Add Support for TeleChat2 Model

Description:

Background: TeleChat2 is an open-source large language model developed by China Telecom's Artificial Intelligence Research Institute. It features various parameter scales and functionalities, including Function Call capabilities.

Modifications:

Model Integration: Added the implementation code for the TeleChat2 model in the vllm/model_executor/models directory.
Model Registration: Registered the TeleChat2 model in the model_registry.py file to enable recognition and utilization within vLLM.
Compatibility Adjustments: Modified the model's forward() method to ensure compatibility with vLLM's architecture.
Testing:

Functional Testing: Conducted inference tests in a local environment to verify the model's proper operation within vLLM.
Performance Evaluation: Assessed inference speed and resource utilization, confirming that performance meets expectations.
Compatibility:

These modifications do not affect vLLM's support for other models.
No new dependencies have been introduced.

Commit:
fix

Commit:
fxii

Isotr0py commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,435 @@
+        bias: bool = False,
+    ) -> None:
+        super().__init__()
+        self.gate_proj = ColumnParallelLinear(
+            input_size=hidden_size,
+            output_size=intermediate_size,
+            bias=False,
+            quant_config=quant_config)
+        
+        self.up_proj = ColumnParallelLinear(
+            input_size=hidden_size,
+            output_size=intermediate_size,
+            bias=False,
+            quant_config=quant_config)
```
We have `MergedColumnParallelLinear` to fuse `gate_up_proj`, so we don't need `torch.cat` in forward runtime.

jeejeelee commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,435 @@
+        self.act_fn = SiluAndMul()
+
+    def forward(self, x):
+        gate_output, _ = self.gate_proj(x)
```
Can we merge linear using `MergedColumnParallelLinear`, see: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L73

jeejeelee commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,435 @@
+        return hidden_states
+
+
+class TeleChat2ForCausalLM(nn.Module, SupportsLoRA):
```
It seems there are no LoRA-related static variables added, it should not inherit `SupportsLoRA`,see: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L439-L453

jeejeelee commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,435 @@
+        cache_config: Optional[CacheConfig] = None,
+        quant_config: Optional[QuantizationConfig] = None,
+        lora_config: Optional[LoRAConfig] = None,
+    ) -> None:
```
Please replace these arguments with `vllm_config`,The same applies to `TeleChat2Model`. You can refer to the `Llama` implementation

Isotr0py commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,435 @@
+        return output
+
+
+class TeleChat2DecoderLayer(nn.Module):
```
Seems that the attention, MLP and decoder layer is basically equivalent to Llama's implementation except submodule naming. Can we re-use Llama's implementation with `WeightsMapper` and `` in `utils.py`?

https://github.com/vllm-project/vllm/blob/e0853b65089b94c9bab9f480970dc73e1e8a0c0d/vllm/model_executor/models/utils.py#L28-L34

https://github.com/vllm-project/vllm/blob/e0853b65089b94c9bab9f480970dc73e1e8a0c0d/vllm/model_executor/models/utils.py#L67-L81

Isotr0py reviewed (commented):
Seems that the model implementation is basically equivalent to Llama except module naming. (Please correct me if I'm wrong)

If so, I think we can simplify the model implementation by mapping weights names.

shunxing12345 commented:
Thank you for your feedback and suggestions!

I’ve noted your points and will make the necessary changes as soon as possible. Additionally, regarding the model implementation, our architecture has some differences from Llama in terms of bias configurations:

- Q and K components include biases, but V does not.
- MLP layers also contain biases.

Because of these differences, Llama cannot directly load our model weights, as it supports only uniform bias configurations (all or none).

I’ll work on these updates and keep you posted! Let me know if you have any further insights or questions.

Commit:
fix

shunxing12345 commented:
Here’s the revised response:

Thank you for your feedback and suggestions!

I’ve noted your points and will make the necessary changes as soon as possible. Additionally, regarding the model implementation, our architecture has some differences from Llama in terms of bias configurations:

- Q and K components include biases, but V does not.
- MLP layers also contain biases.

Because of these differences, Llama cannot directly load our model weights, as it supports only uniform bias configurations (all or none).

I’ve also addressed the points you raised that required modifications. Please feel free to share any further insights or suggestions!

Commit:
fix

Commit:
fix

Commit:
fix

jeejeelee commented:
>Q and K components include biases, but V does not.


> Here’s the revised response:
> 
> Thank you for your feedback and suggestions!
> 
> I’ve noted your points and will make the necessary changes as soon as possible. Additionally, regarding the model implementation, our architecture has some differences from Llama in terms of bias configurations:
> 
> * Q and K components include biases, but V does not.
> * MLP layers also contain biases.
> 
> Because of these differences, Llama cannot directly load our model weights, as it supports only uniform bias configurations (all or none).
> 
> I’ve also addressed the points you raised that required modifications. Please feel free to share any further insights or suggestions!

If these are the only two differences, it might be possible to integrate them following PHI-3's approach, please refer to:  [https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/phi3.py](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/phi3.py)

Isotr0py commented:
> - Q and K components include biases, but V does not.
> - MLP layers also contain biases.

If the only difference is about bias, we can simply set `bias=True` on Llama's Attention and MLP, then handle this in weigts loading by creating full zeros bias for `v_proj` and `MLP` that might not have bias.

shunxing12345 commented:
Thank you for your detailed explanation! In TeleChat2, the bias is set to False for gate_up_proj in the MLP, while it is set to True for down_proj. Additionally, in the Attention module, I previously misspoke. The bias is actually False for qkv_proj, but it is True for o_proj.

If we directly rely on the bias setting in Llama’s MLP and Attention, it might introduce some issues.

Commit:
fix

shunxing12345 commented:
> > Q and K components include biases, but V does not.
> 
> > Here’s the revised response:
> > Thank you for your feedback and suggestions!
> > I’ve noted your points and will make the necessary changes as soon as possible. Additionally, regarding the model implementation, our architecture has some differences from Llama in terms of bias configurations:
> > 
> > * Q and K components include biases, but V does not.
> > * MLP layers also contain biases.
> > 
> > Because of these differences, Llama cannot directly load our model weights, as it supports only uniform bias configurations (all or none).
> > I’ve also addressed the points you raised that required modifications. Please feel free to share any further insights or suggestions!
> 
> If these are the only two differences, it might be possible to integrate this model following PHI-3's approach, please refer to: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/phi3.py

Thank you for your feedback and suggestions!

As per your advice, I have inherited the Llama class and completed the implementation of the TeleChat2 code.

shunxing12345 commented:
> > Q and K components include biases, but V does not.
> 
> > Here’s the revised response:
> > Thank you for your feedback and suggestions!
> > I’ve noted your points and will make the necessary changes as soon as possible. Additionally, regarding the model implementation, our architecture has some differences from Llama in terms of bias configurations:
> > 
> > * Q and K components include biases, but V does not.
> > * MLP layers also contain biases.
> > 
> > Because of these differences, Llama cannot directly load our model weights, as it supports only uniform bias configurations (all or none).
> > I’ve also addressed the points you raised that required modifications. Please feel free to share any further insights or suggestions!
> 
> If these are the only two differences, it might be possible to integrate this model following PHI-3's approach, please refer to: https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/phi3.py

Thank you for your feedback and suggestions!

As per your advice, I have inherited the Llama class and completed the implementation of the TeleChat2 code.

Commit:
fix

Isotr0py commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,285 @@
+        return loaded_params
+    
+
+class TeleChat2ForCausalLM(nn.Module):
+
+    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+        super().__init__()
```
Can we also inherit from `LlamaForCausalLM`? So that we can only override config, `self.model` and `load_weights`.

shunxing12345 replied:
Thank you for your suggestion!

I just tried inheriting from LlamaForCausalLM, but it seems to cause a memory issue where the VRAM usage doubles when loading the model. I suspect this might be related to how the weights or buffers are handled during initialization. Let me know if you have any insights on this, or if there's a workaround we can try!

Thanks again for your input!

Isotr0py replied:
Hmmm, this might because we directly initialize the model on GPU in model loader. Let's keep it as it looks now!

Isotr0py replied:
```suggestion
class TeleChat2ForCausalLM(LlamaForCausalLM):

    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
        super(LlamaForCausalLM, self).__init__()
```

After consideration, I think we can still have a try to inherit from `LlamaForCausalLM` with keeping `__init__` only inherited from `nn.Module`. So that we can avoid implementing `compute_logits` etc duplicatedly.

shunxing12345 replied:
Thank you for your suggestion! I have modified the code accordingly.

Isotr0py commented:
BTW, please address the lint errors by running `bash ./format.sh` and add this model to `docs/source/models/supported_models.rst`. :)

Commit:
fix

shunxing12345 commented:
> BTW, please address the lint errors by running `bash ./format.sh` and add this model to `docs/source/models/supported_models.rst`. :)

I’ve updated the file . I really appreciate your guidance!

shunxing12345 committed:
Update supported_models.rst

shunxing12345 commented:
> BTW, please address the lint errors by running `bash ./format.sh` and add this model to `docs/source/models/supported_models.rst`. :)
![image](https://github.com/user-attachments/assets/8a869cbb-a0ba-41bf-bbbe-764e215c8708)

Hi,

I have already run bash format.sh, and the result is shown in the attached screenshot—everything passed the format check. Additionally, I have updated the supported_models.rst file as well.

Please let me know if there’s anything else needed.

Thank you!

Isotr0py committed:
address lint errors

Signed-off-by: Isotr0py <2037008807@qq.com>

Isotr0py commented:
> I have already run bash format.sh, and the result is shown in the attached screenshot—everything passed the format check.

No worry, I have addressed the lint errors for you. :)

Isotr0py commented on vllm/model_executor/models/registry.py:
```diff
@@ -54,6 +54,7 @@
     "JAISLMHeadModel": ("jais", "JAISLMHeadModel"),
     "JambaForCausalLM": ("jamba", "JambaForCausalLM"),
     "LlamaForCausalLM": ("llama", "LlamaForCausalLM"),
+    "TeleChat2ForCausalLM": ("telechat2", "TeleChat2ForCausalLM"),
```
Can you keep this in alphabet order?

Isotr0py commented:
BTW, can you merge from main branch? The PR branch is pretty out of date, I'm afraid that some tests in CI might fail on too old branch.

shunxing12345 committed:
Merge pull request #1 from vllm-project/main

merge from main

shunxing12345 commented:
> BTW, can you merge from main branch? The PR branch is pretty out of date, I'm afraid that some tests in CI might fail on too old branch.

Thank you for pointing it out! I've already merged the latest code from the main branch. Please let me know if there's anything else you'd like me to check.

Commit:
fix

shunxing12345 commented:
> BTW, please address the lint errors by running `bash ./format.sh` and add this model to `docs/source/models/supported_models.rst`. :)

Thank you for your help!

Commit:
fix

Commit:
fix

Commit:
fix

Isotr0py commented:
Hmmm, seems that the model is not compatible with `torch.compile`, because we're overriding initialized module in inherent:
```
INFO 11-26 09:26:29 config.py:1869] Downcasting torch.float32 to torch.float16.
INFO 11-26 09:26:36 config.py:373] This model supports multiple tasks: {'generate', 'embedding'}. Defaulting to 'generate'.
WARNING 11-26 09:26:36 arg_utils.py:1105] [DEPRECATED] Block manager v1 has been removed, and setting --use-v2-block-manager to True or False has no effect on vLLM behavior. Please remove --use-v2-block-manager in your engine argument. If your use case is not supported by SelfAttnBlockSpaceManager (i.e. block manager v2), please file an issue with detailed information.
INFO 11-26 09:26:36 llm_engine.py:248] Initializing an LLM engine (v0.1.dev3049+g82c2515.d20241019) with config: model='TeleAI/TeleChat2-3B', speculative_config=None, tokenizer='TeleAI/TeleChat2-3B', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=4096, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=TeleAI/TeleChat2-3B, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=True, use_cached_outputs=False, mm_processor_kwargs=None, pooler_config=None,compilation_config=CompilationConfig(level=0, backend='', custom_ops=[], splitting_ops=['vllm.unified_attention', 'vllm.unified_v1_flash_attention'], use_inductor=True, inductor_specialize_for_cudagraph_no_more_than=None, inductor_compile_sizes={}, inductor_compile_config={}, inductor_passes={}, use_cudagraph=False, cudagraph_num_of_warmups=0, cudagraph_capture_sizes=None, cudagraph_copy_inputs=False, pass_config=PassConfig(dump_graph_stages=[], dump_graph_dir=PosixPath('.'), enable_fusion=True, enable_reshape=True), compile_sizes=<function PrivateAttr at 0x7f3047efbac0>, capture_sizes=<function PrivateAttr at 0x7f3047efbac0>, enabled_custom_ops=Counter(), disabled_custom_ops=Counter(), static_forward_context={})
Downloading Model to directory: /root/.cache/modelscope/hub/TeleAI/TeleChat2-3B
WARNING 11-26 09:26:38 tokenizer.py:174] Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
Downloading Model to directory: /root/.cache/modelscope/hub/TeleAI/TeleChat2-3B
Downloading Model to directory: /root/.cache/modelscope/hub/TeleAI/TeleChat2-3B
INFO 11-26 09:26:43 selector.py:217] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
INFO 11-26 09:26:43 selector.py:129] Using XFormers backend.
INFO 11-26 09:26:54 model_runner.py:1100] Starting to load model TeleAI/TeleChat2-3B...
[rank0]: Traceback (most recent call last):
[rank0]:   File "/kaggle/working/vllm/examples/offline_inference_cli.py", line 80, in <module>
[rank0]:     main(args)
[rank0]:   File "/kaggle/working/vllm/examples/offline_inference_cli.py", line 37, in main
[rank0]:     llm = LLM(**asdict(engine_args))
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/utils.py", line 1052, in inner
[rank0]:     return fn(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/entrypoints/llm.py", line 225, in __init__
[rank0]:     self.llm_engine = self.engine_class.from_engine_args(
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/engine/llm_engine.py", line 574, in from_engine_args
[rank0]:     engine = cls(
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/engine/llm_engine.py", line 335, in __init__
[rank0]:     self.model_executor = executor_class(vllm_config=vllm_config, )
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/executor/executor_base.py", line 36, in __init__
[rank0]:     self._init_executor()
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/executor/gpu_executor.py", line 35, in _init_executor
[rank0]:     self.driver_worker.load_model()
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/worker/worker.py", line 153, in load_model
[rank0]:     self.model_runner.load_model()
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/worker/model_runner.py", line 1102, in load_model
[rank0]:     self.model = get_model(vllm_config=self.vllm_config)
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/model_loader/__init__.py", line 12, in get_model
[rank0]:     return loader.load_model(vllm_config=vllm_config)
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/model_loader/loader.py", line 339, in load_model
[rank0]:     model = _initialize_model(vllm_config=vllm_config)
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/model_loader/loader.py", line 106, in _initialize_model
[rank0]:     return model_class(vllm_config=vllm_config, prefix=prefix)
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 209, in __init__
[rank0]:     self.model = TeleChat2Model(vllm_config=vllm_config,
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 139, in __init__
[rank0]:     self.start_layer, self.end_layer, self.layers = make_layers(
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/utils.py", line 510, in make_layers
[rank0]:     [PPMissingLayer() for _ in range(start_layer)] + [
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/utils.py", line 511, in <listcomp>
[rank0]:     maybe_offload_to_cpu(layer_fn(prefix=f"{prefix}.{idx}"))
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 141, in <lambda>
[rank0]:     lambda prefix: TeleChat2DecoderLayer(config=config,
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 110, in __init__
[rank0]:     self.self_attn = TeleChat2Attention(
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 76, in __init__
[rank0]:     super().__init__(config, hidden_size, num_heads, num_kv_heads,
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/llama.py", line 171, in __init__
[rank0]:     self.attn = Attention(
[rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/attention/layer.py", line 110, in __init__
[rank0]:     raise ValueError(f"Duplicate layer name: {prefix}")
[rank0]: ValueError: Duplicate layer name: .attn
```

shunxing12345 commented:
> Hmmm, seems that the model is not compatible with `torch.compile`, because we're overriding initialized module in inherent:
> 
> ```
> INFO 11-26 09:26:29 config.py:1869] Downcasting torch.float32 to torch.float16.
> INFO 11-26 09:26:36 config.py:373] This model supports multiple tasks: {'generate', 'embedding'}. Defaulting to 'generate'.
> WARNING 11-26 09:26:36 arg_utils.py:1105] [DEPRECATED] Block manager v1 has been removed, and setting --use-v2-block-manager to True or False has no effect on vLLM behavior. Please remove --use-v2-block-manager in your engine argument. If your use case is not supported by SelfAttnBlockSpaceManager (i.e. block manager v2), please file an issue with detailed information.
> INFO 11-26 09:26:36 llm_engine.py:248] Initializing an LLM engine (v0.1.dev3049+g82c2515.d20241019) with config: model='TeleAI/TeleChat2-3B', speculative_config=None, tokenizer='TeleAI/TeleChat2-3B', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=4096, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=TeleAI/TeleChat2-3B, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=True, use_cached_outputs=False, mm_processor_kwargs=None, pooler_config=None,compilation_config=CompilationConfig(level=0, backend='', custom_ops=[], splitting_ops=['vllm.unified_attention', 'vllm.unified_v1_flash_attention'], use_inductor=True, inductor_specialize_for_cudagraph_no_more_than=None, inductor_compile_sizes={}, inductor_compile_config={}, inductor_passes={}, use_cudagraph=False, cudagraph_num_of_warmups=0, cudagraph_capture_sizes=None, cudagraph_copy_inputs=False, pass_config=PassConfig(dump_graph_stages=[], dump_graph_dir=PosixPath('.'), enable_fusion=True, enable_reshape=True), compile_sizes=<function PrivateAttr at 0x7f3047efbac0>, capture_sizes=<function PrivateAttr at 0x7f3047efbac0>, enabled_custom_ops=Counter(), disabled_custom_ops=Counter(), static_forward_context={})
> Downloading Model to directory: /root/.cache/modelscope/hub/TeleAI/TeleChat2-3B
> WARNING 11-26 09:26:38 tokenizer.py:174] Using a slow tokenizer. This might cause a significant slowdown. Consider using a fast tokenizer instead.
> Downloading Model to directory: /root/.cache/modelscope/hub/TeleAI/TeleChat2-3B
> Downloading Model to directory: /root/.cache/modelscope/hub/TeleAI/TeleChat2-3B
> INFO 11-26 09:26:43 selector.py:217] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
> INFO 11-26 09:26:43 selector.py:129] Using XFormers backend.
> INFO 11-26 09:26:54 model_runner.py:1100] Starting to load model TeleAI/TeleChat2-3B...
> [rank0]: Traceback (most recent call last):
> [rank0]:   File "/kaggle/working/vllm/examples/offline_inference_cli.py", line 80, in <module>
> [rank0]:     main(args)
> [rank0]:   File "/kaggle/working/vllm/examples/offline_inference_cli.py", line 37, in main
> [rank0]:     llm = LLM(**asdict(engine_args))
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/utils.py", line 1052, in inner
> [rank0]:     return fn(*args, **kwargs)
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/entrypoints/llm.py", line 225, in __init__
> [rank0]:     self.llm_engine = self.engine_class.from_engine_args(
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/engine/llm_engine.py", line 574, in from_engine_args
> [rank0]:     engine = cls(
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/engine/llm_engine.py", line 335, in __init__
> [rank0]:     self.model_executor = executor_class(vllm_config=vllm_config, )
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/executor/executor_base.py", line 36, in __init__
> [rank0]:     self._init_executor()
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/executor/gpu_executor.py", line 35, in _init_executor
> [rank0]:     self.driver_worker.load_model()
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/worker/worker.py", line 153, in load_model
> [rank0]:     self.model_runner.load_model()
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/worker/model_runner.py", line 1102, in load_model
> [rank0]:     self.model = get_model(vllm_config=self.vllm_config)
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/model_loader/__init__.py", line 12, in get_model
> [rank0]:     return loader.load_model(vllm_config=vllm_config)
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/model_loader/loader.py", line 339, in load_model
> [rank0]:     model = _initialize_model(vllm_config=vllm_config)
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/model_loader/loader.py", line 106, in _initialize_model
> [rank0]:     return model_class(vllm_config=vllm_config, prefix=prefix)
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 209, in __init__
> [rank0]:     self.model = TeleChat2Model(vllm_config=vllm_config,
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 139, in __init__
> [rank0]:     self.start_layer, self.end_layer, self.layers = make_layers(
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/utils.py", line 510, in make_layers
> [rank0]:     [PPMissingLayer() for _ in range(start_layer)] + [
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/utils.py", line 511, in <listcomp>
> [rank0]:     maybe_offload_to_cpu(layer_fn(prefix=f"{prefix}.{idx}"))
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 141, in <lambda>
> [rank0]:     lambda prefix: TeleChat2DecoderLayer(config=config,
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 110, in __init__
> [rank0]:     self.self_attn = TeleChat2Attention(
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/telechat2.py", line 76, in __init__
> [rank0]:     super().__init__(config, hidden_size, num_heads, num_kv_heads,
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/model_executor/models/llama.py", line 171, in __init__
> [rank0]:     self.attn = Attention(
> [rank0]:   File "/opt/conda/envs/vllm/lib/python3.10/site-packages/vllm/attention/layer.py", line 110, in __init__
> [rank0]:     raise ValueError(f"Duplicate layer name: {prefix}")
> [rank0]: ValueError: Duplicate layer name: .attn
> ```
> 
> Will take a look tonight to see how we can do some refactor on curret implementation. :(

yeah... I was wondering if this issue could be related to recent changes in LLaMA, as the same code works perfectly fine with vLLM 0.6.4.post1. If that’s the case, I’m a bit concerned that similar compatibility issues might arise again in the future if LLaMA undergoes further updates after we make the necessary adjustments. Do you think there’s a way to address this in a more stable manner moving forward? I’d love to hear your thoughts.

shunxing12345 committed:
Merge pull request #2 from vllm-project/main

merge from main

Commit:
fix

Commit:
Merge branch 'main' of https://github.com/shunxing12345/vllm

shunxing12345 commented:
> @shunxing12345 I think a possible way is pruning the bias from a LlamaModel.
> 
> I have drafted a refactored implementation for `telechat2` on my fork: https://github.com/vllm-project/vllm/blob/4ac891f8ccfc2ef1260cc9023fbcb62d1f876c26/vllm/model_executor/models/telechat2.py.
> 
> I have tested it on `Telechat2-3B` and the 2x VRAM issue should be resolved as well with this implementation. But I haven't handled `qkv_bias` etc config for bias prune, feel free to copy to this PR and make any modifications that you think necessary :)

![image](https://github.com/user-attachments/assets/9b3f6f0c-894a-4e77-af6d-9afca4102f29)


Thank you for your assistance! I have tried your code, but when using vLLM to serve the 35B and 115B models, the process gets stuck at this stage (as shown in the screenshot) for more than 10 minutes without any response. Could you please provide guidance on how to resolve this issue? Your help would be greatly appreciated!

Isotr0py commented:
> I have tried your code, but when using vLLM to serve the 35B and 115B models, the process gets stuck at this stage (as shown in the screenshot) for more than 10 minutes without any response.

Perhaps you can try offline inference instead. Just run:
```shell
VLLM_USE_MODELSCOPE=True python examples/offline_inference_cli.py --model TeleAI/TeleChat2-3B --max-model-len 4096 --trust-remote-code
```

shunxing12345 commented:
> > I have tried your code, but when using vLLM to serve the 35B and 115B models, the process gets stuck at this stage (as shown in the screenshot) for more than 10 minutes without any response.
> 
> Perhaps you can try offline inference instead. Just run:
> ```shell
> VLLM_USE_MODELSCOPE=True python examples/offline_inference_cli.py --model TeleAI/TeleChat2-3B --max-model-len 4096 --trust-remote-code
> ```

Thank you so much for your guidance and help! Apologies for my earlier mistake—now I have successfully gotten TeleChat2 to run across all sizes. Your assistance has been incredibly helpful. Thank you again!😊

Isotr0py commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,139 @@
+        vllm_config.model_config.hf_config.mlp_bias = True
+        super().__init__(vllm_config=vllm_config, prefix=prefix)
+        # 2. Remove the bias from the qkv_proj and gate_up_proj based on config
+        # FIXME: Handle qkv_bias etc
```
```suggestion
        # Telechat2's gate_up_proj and qkv_proj don't have bias
        # see: https://github.com/vllm-project/vllm/pull/10311#issuecomment-2490297566
```

shunxing12345 replied:
Thank you for the review and approval!

Isotr0py reviewed (approved):
LGTM! Thanks for adding this model!

Commit:
fix

Isotr0py committed:
fix registry test

Signed-off-by: Isotr0py <2037008807@qq.com>

DarkLight1337 commented on tests/models/registry.py:
```diff
@@ -115,6 +115,9 @@ class _HfExamplesInfo:
     "StableLmForCausalLM": _HfExamplesInfo("stabilityai/stablelm-3b-4e1t"),
     "Starcoder2ForCausalLM": _HfExamplesInfo("bigcode/starcoder2-3b"),
     "SolarForCausalLM": _HfExamplesInfo("upstage/solar-pro-preview-instruct"),
+    "TeleChat2ForCausalLM": _HfExamplesInfo("Tele-AI/TeleChat2-115B",
```
```suggestion
    "TeleChat2ForCausalLM": _HfExamplesInfo("Tele-AI/TeleChat2-115B",
```

To ease testing, please use a smaller model here.

Isotr0py replied:
@shunxing12345 Seems that `TeleChat2-3B` is only released on modelscope. Can you upload the `TeleChat2-3B` to HF for test?

shunxing12345 replied:
@Isotr0py Hi, I've uploaded it to HF.  https://huggingface.co/Tele-AI/TeleChat2-3B

DarkLight1337 commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,147 @@
+class TeleChat2ForCausalLM(LlamaForCausalLM):
+
+    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
+        super(LlamaForCausalLM, self).__init__()
```
```suggestion
        super().__init__()
```

Isotr0py replied:
Hmmm, this model uses a `TeleChat2Model` instead of `LlamaModel`, which is different from Llama. If we directly inherit from `LlamaForCausalLM`, this will break the model compatibility with `torch.compile` because of duplicate initialization.

See: https://github.com/vllm-project/vllm/pull/10311#issuecomment-2500103544

Isotr0py commented on docs/source/models/supported_models.rst:
```diff
@@ -309,6 +309,11 @@ Text Generation
+  * - :code:`TeleChat2ForCausalLM`
+    - TeleChat2
+    - :code:`TeleAI/TeleChat2-3B`, :code:`TeleAI/TeleChat2-7B`, :code:`TeleAI/TeleChat2-35B`, etc.
+    - 
```
Oh, we need to handle `lm_head` with PP and LoRA config for `TeleChat2ForCausalLM` in intialization, because we overwrite the `__init__` from `LlamaForCausalLM`.

Let me make a quick fix for it.

Isotr0py replied:
`TeleChat2ForCausalLM` should also support PP now.

DarkLight1337 commented on vllm/model_executor/models/telechat2.py:
```diff
@@ -0,0 +1,147 @@
+        self.model = TeleChat2Model(vllm_config=vllm_config,
+                                    prefix=maybe_prefix(prefix, "model"))
+
+        self.lm_head = ParallelLMHead(config.vocab_size,
+                                      config.hidden_size,
+                                      bias=False,
+                                      quant_config=quant_config,
+                                      prefix=maybe_prefix(prefix, "lm_head"))
+        self.logits_processor = LogitsProcessor(config.vocab_size)
+        self.sampler = Sampler()
+        self._pooler = Pooler.from_config_with_defaults(
+            pooler_config,
+            pooling_type=PoolingType.STEP,
+            normalize=False,
+            softmax=False)
```
If these are the same as `LlamaForCausalLM`, we can skip them.

Isotr0py replied:
I think the difference is only the `model`, which is small. Perhaps we can add a `make_model` method for `llama` to make model backbone can be easily replaced?

Isotr0py committed:
cleanup

Signed-off-by: Isotr0py <2037008807@qq.com>

Isotr0py committed:
make telechat2 config compatiable with Llama

Signed-off-by: Isotr0py <2037008807@qq.com>

Isotr0py committed:
code format

Signed-off-by: Isotr0py <2037008807@qq.com>

Isotr0py committed:
make format happy

Signed-off-by: Isotr0py <2037008807@qq.com>

Isotr0py committed:
fix PP and use 3B for test

Signed-off-by: Isotr0py <2037008807@qq.com>
